Mathlib/RingTheory/DedekindDomain/Instances.lean

```diff
@@ -0,0 +1,135 @@
+/-
+Copyright (c) 2025 Riccardo Brasca. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Riccardo Brasca
+-/
+
+import Mathlib.Algebra.Lie.OfAssociative
+import Mathlib.Order.CompletePartialOrder
+import Mathlib.RingTheory.DedekindDomain.PID
+import Mathlib.FieldTheory.Separable
+
+/-!
+# Instances for Dedekind domains
+This file contains various instances to work with localization of a ring extension.
+
+A very common situation in number theory is to have an extension of (say) Dedekind domains `R` and
+`S`, and to prove a property of this extension is useful to consider the localization `Rₚ` of `R` at
+`P`, a prime ideal of `R`. One also works with the corresponding localization `Sₚ` of `S` and the
+fraction fields `K` and `L` of `R` and `S`. In this situation there are many compatible algebra
+structures and various properties of the rings involved. This file contains a collection of such
+instances.
+
+## Implementation details
+In general one wants all the results below for any algebra satisfyng `IsLocalization`, but those
+cannot be instances (since Lean has no way of guessing the submonoid). Having the instances in the
+special case of *the* localization at a prime ideal is useful in working with Dedekind domains.
+
+-/
+
+open nonZeroDivisors IsLocalization Algebra IsFractionRing IsScalarTower
+
+attribute [local instance] FractionRing.liftAlgebra
+
+variable {R : Type*} (S : Type*) [CommRing R] [CommRing S] [IsDomain R] [IsDomain S] [Algebra R S]
+
+local notation3 "K" => FractionRing R
+local notation3 "L" => FractionRing S
+section
+
+theorem map_le_nonZeroDivisors_of_faithfulSMul {A : Type*} (B : Type*) [CommSemiring A]
+    [CommSemiring B] [Algebra A B] [NoZeroDivisors B] [FaithfulSMul A B] {S : Submonoid A}
+    (hS : S ≤ A⁰) : algebraMapSubmonoid B S ≤ B⁰ :=
+  map_le_nonZeroDivisors_of_injective _ (FaithfulSMul.algebraMap_injective A B) hS
```

Comment:
This lemma strikes me as the straightforward combination of two simpler lemmas. Do we really need it?

Reply:
I think it is convenient to keep it: it is used 4 times, and replacing it with its proof is a little cumbersome (Lean needs help with some instance).